### DISCUSSION TITLE
Netlab integration with Cisco CML

### DISCUSSION THREAD

**Discussion by danohn** [ipspace/netlab#1528] [Ideas]
Cisco CML currently provides a REST API that allows for the creation and deletion of labs. Labs are defined in YAML format which defines the links as well as the config itself. This seems to be a perfect use case for Netlab where Netlab could create the lab, links and config. Is there a possibility of this being on the roadmap?

**Comment by Astrotokii** [comment]
Could you pull the snapshot YAML file from your VM and upload to CML? Or are you recommending a plugin for netlab?

**Comment by DanPartelly** [comment]
There is no such feature in the netlab roadmap. 

What would be the value of this one ? Once I have a lab up and running, why would I want to transfer it to CML and work it in CML instead of just working it?  Its already running. 

You could write a 3rd party tool to parse netlab data structures + generated configs and transform them into what CML wants to digest.   

> Cisco CML currently provides a REST API that allows for the creation and deletion of labs. Labs are defined in YAML format which defines the links as well as the config itself. This seems to be a perfect use case for Netlab where Netlab could create the lab, links and config. Is there a possibility of this being on the roadmap?

**Reply by ipspace** [reply]
> What would be the value of this one ? Once I have a lab up and running, why would I want to transfer it to CML and work it in CML instead of just working it? Its already running.

You wouldn't have to pull the Cisco images from the CML ISO file and build the Vagrant boxes or containers. Also, with CML running "somewhere", you could run netlab/ansible on any platform, including an Apple Silicon laptop.

**Comment by ipspace** [comment]
If someone has a running CML and enough Python chops, a CML provider would be quite welcome. It would also nicely answer the occasional question of "why should I use netlab instead of CML" 😉

Such a provider would:

- Create a YAML file during the **netlab create** processing
- Use the CML API to upload that YAML file in the **start_lab** hook (invoked by **netlab up**)
- If needed, fetch the management IP addresses of the CML VMs and update whatever needs to be updated so that the rest of **netlab** commands would know how to access the devices. Alternatively, if the VMs can be reached under different ports on the CML server's IP address, we could use a hack similar to what we used with VirtualBox.
- **netlab initial** (invoked as a command or as part of **netlab up**) would then deploy the netlab-generated configuration.

This approach might work better than building the whole configuration and storing it into a YAML file. **netlab** never generates a complete device configuration; every module-specific device template (for example, `bgp/ios.j2`) adds the stuff needed to deploy that module.

**Reply by danohn** [reply]
Thank you for the reply. For some historical context, CML's (previously known as VIRL 2) predecessor, VIRL 1, used a tool called [AutoNetkit](https://github.com/sk2/autonetkit). The role of AutoNetkit was to bootstrap all of the configuration such as P2P IP's, Loopback IP's, OSPF, MPLS, BGP etc. Unfortunately, CML has dropped support for AutoNetkit.

While there is another project that aims to solve this called [CMLNetkit](https://github.com/WojciechowskiPiotr/CMLNetKit), I personally prefer Netlab's approach to bootstrapping the lab.
